cupy/_math/sumprod.py

```diff
@@ -507,7 +507,66 @@ def gradient(f, *varargs, axis=None, edge_order=1):
         return outvals
 
 
-# TODO(okuta): Implement ediff1d
+def ediff1d(arr, to_end=None, to_begin=None):
+    """
+    Calculates the difference between consecutive elements of an array.
+
+    Args:
+        arr (cupy.ndarray): Input array.
+        to_end (cupy.ndarray, optional): Numbers to append at the end
+            of the returend differences.
+        to_begin (cupy.ndarray, optional): Numbers to prepend at the
+            beginning of the returned differences.
+
+    Returns:
+        cupy.ndarray: New array consisting differences among succeeding
+        elements.
+
+    .. seealso:: :func:`numpy.ediff1d`
+    """
+    # to flattened array.
+    arr = cupy.array(arr).ravel()
```

Comment:
`cupy.array()` works as same as `cupy.asanyarray()`, and it can also bring host-to-device sync. How about the following (also `to_begin` and `to_end`)?
```suggestion
    if not isinstance(arr, cupy.ndarray):
        raise TypeError('`arr` should be of type cupy.ndarray')
    arr = arr.ravel()
```

Reply:
I think it's a more thoughtful approach to do so. Thanks for it! Just curious to know why do we need to avoid host-to-device sync here?

Also, I tried for `to_begin` and `to_end` too. I got the following results.
#### Code
```
if to_begin is None:
        l_begin = 0
    else:
        if not isinstance(to_begin, numpy.ndarray):
            raise TypeError('`to_begin` should be of type numpy.ndarray')
        if not cupy.can_cast(to_begin, dtype_req, casting="same_kind"):
            raise TypeError("dtype of `to_begin` must be compatible "
                            "with input `arr` under the `same_kind` rule.")

        to_begin = to_begin.ravel()
        l_begin = len(to_begin)
```
#### Result
```
tests/cupy_tests/math_tests/test_sumprod.py:1066: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cupy/_math/sumprod.py:578: in ediff1d
    result[:l_begin] = to_begin
cupy/_core/core.pyx:1498: in cupy._core.core.ndarray.__setitem__
    _indexing._ndarray_setitem(self, slices, value)
cupy/_core/_routines_indexing.pyx:51: in cupy._core._routines_indexing._ndarray_setitem
    _scatter_op(self, slices, value, 'update')
cupy/_core/_routines_indexing.pyx:987: in cupy._core._routines_indexing._scatter_op
    y.fill(value)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise ValueError(
E   AssertionError: Only cupy raises error
E   
E   Traceback (most recent call last):
E     File "/home/khushi/Documents/cupy/cupy/testing/_loops.py", line 43, in _call_func
E       result = impl(*args, **kw)
E     File "/home/khushi/Documents/cupy/tests/cupy_tests/math_tests/test_sumprod.py", line 1066, in test_ediff1d_to_begin1
E       return xp.ediff1d(a, to_begin=numpy.array([0], dtype=dtype))
E     File "/home/khushi/Documents/cupy/cupy/_math/sumprod.py", line 578, in ediff1d
E       result[:l_begin] = to_begin
E     File "cupy/_core/core.pyx", line 1498, in cupy._core.core.ndarray.__setitem__
E       _indexing._ndarray_setitem(self, slices, value)
E     File "cupy/_core/_routines_indexing.pyx", line 51, in cupy._core._routines_indexing._ndarray_setitem
E       _scatter_op(self, slices, value, 'update')
E     File "cupy/_core/_routines_indexing.pyx", line 987, in cupy._core._routines_indexing._scatter_op
E       y.fill(value)
E     File "cupy/_core/core.pyx", line 676, in cupy._core.core.ndarray.fill
E       raise ValueError(
E   ValueError: non-scalar numpy.ndarray cannot be used to fill

cupy/_core/core.pyx:676: AssertionError
----------------------------------------- Captured stdout call -----------------------------------------
dtype is <class 'numpy.float64'>
```
#### My viewpoints
- We cannot use `cupy.ndarray` type in `to_begin` and `to_end` to `isinstance` for the reason that CuPy's `ndarray` does not support `subclassing` as of now.
- For `numpy.ndarray`, it throws the mentioned error as we cannot access the CPU memory. Credits: [StackOverflow](https://stackoverflow.com/questions/68634058)
- Since, we are passing `numpy.ndarray` as input for `to_begin` and `to_end`, which runs on CPU. We need to transfer the data to GPU. Henceforth, we need host-to-device sync.

Do these sound good to you? Will you please let me know if I am getting something wrong?

Thanks a lot for reviewing the PR and for all valuable suggestions (:

cc: @takagi

Reply:
Host-to-device synchronization blocks the execution until CUDA kernels that have been launched before calling the sync complete their work. If `cupy.ediff1d` implicitly synchronized and the user was not aware of it, the performance of the system could be unintentionally incurred. So we have a design policy that we require users to explicitly manage host-to-device synchronization.

For checking `to_begin` and `to_end`, I don't think we need CuPy to support subclassing and the following would be fine. Can I have why you think so?

```Python
if not isinstance(to_begin, cupy.ndarray):
    raise TypeError('`to_begin` should be of type cupy.ndarray')
```

Reply:
>  If cupy.ediff1d implicitly synchronized and the user was not aware of it, the performance of the system could be unintentionally incurred. So we have a design policy that we require users to explicitly manage host-to-device synchronization.

This is fascinating!

Reply:
> For checking to_begin and to_end, I don't think we need CuPy to support subclassing and the following would be fine. Can I have why you think so?

Yes, you are absolutely correct. They ain't related to subclassing. I was wrong. I revisited the concept and felt so. Thanks for it!

cc: @takagi